lib/src/main/kotlin/net/dean/jraw/http/HttpRequest.kt

```diff
@@ -169,9 +169,18 @@ class HttpRequest internal constructor(
          * Sets this Builder's path to the given Endpoint's path and changes the host to `oauth.reddit.com`. This method
          * does not change the HTTP method.
          */
-        fun endpoint(e: Endpoint, vararg pathParams: String): Builder {
+        fun endpoint(e: Endpoint, vararg pathParams: String?): Builder {
             this.host = "oauth.reddit.com"
-            return path(e.path, *pathParams)
+            return if (e.path.startsWith(Endpoint.Constant.OPTIONAL_SUBREDDIT)) {
+                if (pathParams.isEmpty())
+                    throw IllegalArgumentException("Expected at least one path argument for endpoint with optional subreddit")
+                val optionalPath = pathParams[0]?.let { "/r/${JrawUtils.urlEncode(it)}" } ?: ""
+                val requiredPath = e.path.substringAfter(Endpoint.Constant.OPTIONAL_SUBREDDIT)
+                val requiredPathParams = pathParams.drop(1).toTypedArray().requireNoNulls()
```

Comment:
Sure, this is a viable solution. On the other hand, it has its own cost by adding some ambiguity. For example, right now every `Endpoint` has a certain number of arguments it requires, and if the number of arguments doesn't match exactly, an exception is thrown:
```java
        // Taken from HttpRequest.substitutePathParameters(path: String, positionalArgs: List<String>)
        val pathParams = parsePathParams(path)
        if (pathParams.size != positionalArgs.size) {
            // Different amount of parameters
            throw IllegalArgumentException(
                "URL parameter size mismatch. Expecting ${pathParams.size}, got ${positionalArgs.size}")
        }
```
By making `endpoint()` smarter we get into situation where initial call could contain 1 path argument more than the endpoint requires while not having the optional subreddit argument. Such call would pass the checks and would ultimatelly fail as the path parameters would get messed up.

For example:
```java
    it.endpoint(Endpoint.GET_COMMENTS_ARTICLE, articleId, contextSizeString)
```
would result in 
`GET /r/{articleId}/comments/{contextSizeString}`
 and the library wouldn't throw any exceptions, whilist the user might have meant to sent 
`GET /comments/{articleId}/?context={contextSizeString}`
which is not a valid endpoint and exception should have been thrown.

Just wanted to make that trade-off clear, I am still not sure which method should be preferred. 

Passing nulls is ugly, allowing nullable strings where we throw an exception if any (except the optional first) string is null isn't great as well. But it is an explicit fail-fast mechanism and it won't make any weird API requests which may actually succeed and are hell to debug (and fix consequences if some action was taken because of that wrong request).

I have an idea of a bigger change that would eliminate passing nulls while not allowing any ambiguity, but it would be a bigger change and I want to hear your opinion on that ambiguity point.

Reply:
I think you're right, the less ambiguity the better. What's your idea?

Reply:
The idea is to split the current `endpoint()` implementation into 3 methods: `endpoint()` and depending on whether the optional subreddit is supplied `subreddit(nonNullSubredditName))` or `noSubreddit()` (or `noOptionalSubreddit()` or `frontPage()`, not sure which name fits the best).

This leaves the `endpoint()` method without any ambiguity on the number of supplied String varargs.

And to make the choice explicit (whether the optional subreddit is needed) if the `Endpoint` contains the optional subreddit the user is required to invoke either of these methods at least once. This would be controlled by 2 new inner variables in `HttpRequest`, namely `subreddit: String?` and `optionalSubredditWasSet: Boolean`.

So we have now 2 boolean parameters to construct a table of "what happens when you build the request":

     1. Does the Endpoint (if any) have an optional subreddit?
     2. Has the user invoked one of the two methods regarding the optional subreddit?


1. `false, false` -> request builds successfully
2. `false, true` -> throw exception "you tried to set optional subreddit, but the endpoint doesn't have one"
3. `true, false` -> throw exception "the endpoint has an optional subreddit, but you haven't invoked neither `subreddit(${subredditName})` or `noOptionalSubreddit()`
4. `true, true` -> request builds successfully


 
So now instead of 
```java
it.endpoint(Endpoint.GET_COMMENTS_ARTICLE, null, id)
```
we make the user do next (case #&#xfeff;4):
```java
it.endpoint(Endpoint.GET_COMMENTS_ARTICLE, id)
it.noSubreddit()
```
*(the order of the statements doesn't matter)*

or if the user wants to provide a subreddit (case #&#xfeff;4):
```java
it.subreddit(subredditName)
it.endpoint(Endpoint.GET_WIKI_PAGE, name)
```

And in the case of endpoints without an optional subreddit, it is the same as was always before (case #&#xfeff;1):
```java
it.endpoint(Endpoint.GET_ME_KARMA)
```

Now to clear up why we should throw an exception in the other two cases. 

Case #&#xfeff;2 - user wants to narrow down the endpoint results to a single subreddit, but the request doesn't support that. Not only this would be an illegal argument, but it also would make the statement to set a subreddit just a blank action that doesn't make any difference. We may also have a boolean variable or some method of checking that the request was created using `endpoint()` method and not `url()` or `host()` or `path()` and using them would also result in this case. 

Case #&#xfeff;3 - the endpoint supports an optional subreddit but the user hasn't made the choice. This is the most questionable case, and my opinion is that making this case use a default choice (request without any subreddits) makes it easier for a developer to either forget to set the subreddit or just miss that this endpoint supports that. By making the choice mandatory, the developer makes the choice explicitly.

Upsides: No passing nulls, no ambiguity, fail before the request is created, descriptive error messages. 

Downsides: The `HttpRequest` class gets more complex and more Reddit-centric

Reply:
I think this is a step in the right direction. I don't see `HttpRequest` being used outside of JRAW when someone can use OkHttp directly instead, so being more reddit-centric isn't a problem.

Can we make this a separate PR? This one is pretty big already.

Reply:
Yep, let's leave the null solution for now, after we get done with this PR I will start working on a new one which would consist of implementing the above mentioned idea